src/components/expression/expression-renderer.js

```diff
@@ -459,58 +604,67 @@ genotet.ExpressionRenderer.prototype.drawGeneProfiles_ = function() {
   if (!this.data.options.showProfiles) {
     this.svgProfile_.selectAll('*').attr('display', 'none');
     return;
-  }
-  else {
+  } else {
     if (!this.data.options.showGeneLabels) {
-      this.PROFILE_MARGINS_.LEFT = this.DEFAULT_PROFILE_MARGIN;
-    }
-    else {
-      this.PROFILE_MARGINS_.LEFT = this.geneLabelWidth_ + this.LABEL_MARGIN_;
+      this.profileMargins_.left = this.DEFAULT_PROFILE_MARGIN;
+    } else {
+      this.profileMargins_.left = this.geneLabelWidth_ + this.LABEL_MARGIN;
     }
     this.svgProfile_.selectAll('*').attr('display', 'inline');
   }
 
   var heatmapData = this.data.matrix;
-  this.svgProfile_.selectAll('g').remove();
   this.svgProfile_.attr('width', this.canvasWidth_);
 
+  this.svgLegend_
+    .attr('transform', genotet.utils.getTransform([
+      this.profileMargins_.left,
+      this.DEFAULT_PROFILE_LEGEND_MARGIN
+    ]));
+  var legendHeight = this.DEFAULT_PROFILE_LEGEND_HEIGHT -
+    this.DEFAULT_PROFILE_LEGEND_MARGIN;
+
+  var i = 0;
```

Comment:
This block of code is a pre-processing step that does not need to be repeated on redraw. You can compute this whenever the data is changed.

Besides, indexOf is inefficient. A better way could be
```js
// In utils.js
/**
 * Converts an array of string or object with string keys to a set with the strings
 * as keys. The values will be set to all true.
 * @param {!Array<string>|!Object<*>} collection
 * @return {!Object<boolean>}
 * @private
 */
genotet.utils.keySet_ = function(collection) {
  var obj = {};
  if (collection instanceof Array) {
    collection.forEach(function (element) {
      obj[element] = true;
    });
  } else {
    for (var key in collection) {
      obj[key] = true;
    }
  }
  return obj;
};

// Here
var geneNames = genotet.utils.keySet(heatmapData.geneNames);
this.data.profiles = this.data.profiles.filter(function(profile) {
  return profile.geneName in geneNames;
});
```